File: granola_share/pipeline.py
"""Turn queued notes into filed ones: write our summary from the transcript, sort, save.

Runs as a background thread inside `granola-share run`, so uploads from your laptop return right
away even when a big model takes a few minutes per lecture.
"""

from __future__ import annotations

import dataclasses
import json
import threading
import time
import traceback
from pathlib import Path

from .classify import classify, classify_with_ollama, ollama_chat
from .config import Config
from .store import Classification, Store
from .summarize import summarize_transcript, wants_summary


class Pipeline:
    def __init__(self, cfg: Config, store: Store, *, chat=None, summarize=None, log=print):
        self.cfg = cfg
        self.store = store
        self.chat = chat  # classify's Ollama call (tests swap it)
        self.summarize = summarize or summarize_transcript
        self.log = log
        self._wake = threading.Event()
        self.current: str | None = None  # id of the note being written right now

    def process(self, row) -> Path | None:
        m = self.store.meeting(row)
        summary, model, error = "", "", ""
        if wants_summary(m, self.cfg):
            model = self.cfg.effective_summary_model
            started = time.time()
            try:
                summary = self.summarize(m, self.cfg)
                self.log(f"[pipeline] summarized '{m.title}' with {model} in {time.time() - started:.0f}s")
            except Exception as e:
                error = f"Summary with {model} failed: {e}"
                self.log(f"[pipeline] {error}")
                summary, model = "", ""
        if row["classified_by"] == "human" and row["class_name"]:
            # A person filed it: rewriting the summary never undoes that.
            c = Classification(row["class_name"], 1.0, "human", row["lecture_title"] or "",
                               json.loads(row["topics"] or "[]"))
        else:
            # Sort on our summary when we have one: it is cleaner than the raw transcript.
            basis = dataclasses.replace(m, notes_markdown=summary) if summary else m
            c = classify(basis, self.cfg, self.chat)
            if c.by in ("folder", "rules") and self.cfg.ollama_enabled and self.cfg.classes:
                # The folder or title picked the class; still let the model name the lecture and tag topics.
                described = classify_with_ollama(basis, self.cfg.classes, self.cfg, self.chat or ollama_chat)
                if described:
                    c.lecture_title = described.lecture_title or c.lecture_title
                    c.topics = described.topics
        path = self.store.finish(row, m, c, summary, model, error, keep_granola=self.cfg.keep_granola_notes)
        if path is None:
            self.log(f"[pipeline] '{m.title}' changed while it was being written; not saving the old result")
        else:
            self.log(f"[pipeline] filed '{m.title}' → {c.class_name} ({c.by} {c.confidence:.2f})")
        return path

    def run_pending(self, stop: threading.Event | None = None) -> int:
        done = 0
        while not (stop and stop.is_set()):
            row = self.store.claim_next()
            if row is None:
                break
            self.current = row["id"]
            try:
                self.process(row)
                done += 1
            except Exception as e:
                self.store.mark_failed(row["id"], str(e))
                self.log(f"[pipeline] could not file {row['id']}: {e}\n{traceback.format_exc()}")
            finally:
                self.current = None
        return done

    def wake(self) -> None:
        self._wake.set()

    def run_forever(self, stop: threading.Event) -> None:
        n = self.store.reset_working()
        if n:
            self.log(f"[pipeline] resuming {n} note(s) interrupted by a restart")
        while not stop.is_set():
            try:
                self.run_pending(stop)
            except Exception as e:
                self.log(f"[pipeline] error: {e}")
            self._wake.wait(30)
            self._wake.clear()

    def start(self, stop: threading.Event) -> threading.Thread:
        t = threading.Thread(target=self.run_forever, args=(stop,), name="pipeline", daemon=True)
        t.start()
        return t

File: engine/src/StudyStash.Core/Pipeline.cs
using System.Diagnostics;

namespace StudyStash.Core;

/// <summary>
/// Turns queued lectures into filed ones: writes our notes from the transcript, sorts, saves. Runs in the
/// background of the library, so an upload from the laptop returns at once even when a big model takes minutes.
/// </summary>
public sealed class Pipeline(Config cfg, Store store, SortChatFn? chat = null,
    Func<Meeting, Config, Task<string>>? summarize = null, Action<string>? log = null, Func<string>? notesModel = null)
{
    readonly Func<Meeting, Config, Task<string>> summarize = summarize ?? ((m, c) => Summarize.SummarizeTranscriptAsync(m, c));
    readonly Action<string> log = log ?? Console.WriteLine;
    readonly SemaphoreSlim wake = new(0, 1);
    volatile string? current;

    public Config Cfg { get; } = cfg;

    /// <summary>The id of the lecture being written right now.</summary>
    public string? Current => current;

    public async Task<string?> ProcessAsync(NoteRow row)
    {
        var m = store.Meeting(row);
        string summary = "", model = "", error = "";
        if (Summarize.WantsSummary(m, Cfg))
        {
            model = notesModel?.Invoke() ?? Cfg.EffectiveSummaryModel; // what writes the notes, as the note records it
            var watch = Stopwatch.StartNew();
            try
            {
                summary = await summarize(m, Cfg);
                log($"[pipeline] summarized '{m.Title}' with {model} in {watch.Elapsed.TotalSeconds:0}s");
            }
            catch (Exception e)
            {
                error = $"Summary with {model} failed: {e.Message}";
                log($"[pipeline] {error}");
                summary = model = "";
            }
        }
        Classification c;
        if (row.ClassifiedBy == "human" && !string.IsNullOrEmpty(row.ClassName))
        {
            // A person filed it: writing the notes again never undoes that.
            c = new Classification(row.ClassName, 1.0, "human", row.LectureTitle ?? "", JsonTopics(row.Topics));
        }
        else
        {
            // Sort on our notes when we have them: they are cleaner than the raw transcript.
            var basis = summary.Length > 0 ? m.WithNotes(summary) : m;
            c = await Classify.ClassifyAsync(basis, Cfg, chat);
            if (c.By is "folder" or "rules" && Cfg.OllamaEnabled && Cfg.Classes.Count > 0)
            {
                // The folder or title picked the class; still let the model name the lecture and tag its topics.
                var described = await Classify.WithOllamaAsync(basis, Cfg.Classes, Cfg, chat);
                if (described is not null)
                {
                    if (described.LectureTitle.Length > 0) c.LectureTitle = described.LectureTitle;
                    c.Topics = described.Topics;
                }
            }
        }
        string? path = store.Finish(row, m, c, summary, model, error);
        log(path is null
            ? $"[pipeline] '{m.Title}' changed while it was being written; not saving the old result"
            : $"[pipeline] filed '{m.Title}' \u2192 {c.ClassName} ({c.By} {Py.FormatFixed(c.Confidence, 2)})");
        return path;
    }

    static List<string> JsonTopics(string? json) =>
        (System.Text.Json.Nodes.JsonNode.Parse(string.IsNullOrEmpty(json) ? "[]" : json) as System.Text.Json.Nodes.JsonArray
         ?? new System.Text.Json.Nodes.JsonArray()).Select(Py.Str).ToList();

    public async Task<int> RunPendingAsync(CancellationToken stop = default)
    {
        int done = 0;
        while (!stop.IsCancellationRequested)
        {
            var row = store.ClaimNext();
            if (row is null) break;
            current = row.Id;
            try
            {
                await ProcessAsync(row);
                done++;
            }
            catch (Exception e)
            {
                store.MarkFailed(row.Id, e.Message);
                log($"[pipeline] could not file {row.Id}: {e.Message}\n{e}");
            }
            finally
            {
                current = null;
            }
        }
        return done;
    }

    /// <summary>A new lecture arrived: look at the queue now instead of within 30 seconds.</summary>
    public void Wake()
    {
        try
        {
            wake.Release();
        }
        catch (SemaphoreFullException)
        {
            // already woken
        }
    }

    public async Task RunForeverAsync(CancellationToken stop)
    {
        int n = store.ResetWorking();
        if (n > 0) log($"[pipeline] resuming {n} note(s) interrupted by a restart");
        while (!stop.IsCancellationRequested)
        {
            try
            {
                await RunPendingAsync(stop);
            }
            catch (Exception e)
            {
                log($"[pipeline] error: {e.Message}");
            }
            try
            {
                await wake.WaitAsync(TimeSpan.FromSeconds(30), stop);
            }
            catch (OperationCanceledException)
            {
                break;
            }
        }
    }

    public Task Start(CancellationToken stop) => Task.Run(() => RunForeverAsync(stop), CancellationToken.None);
}
